File: Array/MaxChunksToMakeSortedII/MaxChunksToMakeSortedII.go
/**
 * @input A : Integer array
 * 
 * @Output Integer
 */

 func min(a int, b int) (int) {
    if a < b {
        return a
    } else {
        return b
    }
}
func max(a int, b int) (int) {
    if a > b {
        return a
    } else {
        return b
    }
}
func solve(A []int )  (int) {
    r_min := make([]int,len(A))
    r_min[len(A)-1] = A[len(A)-1]
    for i:=len(A)-2;i>=0;i-- {
        r_min[i]=min(r_min[i+1],A[i+1])
    }
    l_max := A[0]
    res := 0
    for i:=0;i<len(A);i++ {
        l_max=max(l_max,A[i])
        if i == len(A)-1 || l_max <= r_min[i] {
            res++
        }
    }
    return res
}

File: Array/MaxChunksToMakeSortedII/MaxChunksToMakeSortedII.py
class Solution:
    # @param A : list of integers
    # @return an integer
    def solve(self, A):
        r_min = [A[-1]]*len(A)
        for i in range(len(A)-2,-1,-1):
            r_min[i]=min(r_min[i+1],A[i+1])
        l_max=A[0]
        res = 0
        for i in range(len(A)):
            l_max=max(l_max,A[i])
            if i == len(A)-1 or l_max <= r_min[i]:
                res+=1
        return res